File: app/src/main/java/io/github/xalrk/nudge/domain/ImportExport.kt
package io.github.xalrk.nudge.domain

import io.github.xalrk.nudge.data.Kind
import io.github.xalrk.nudge.data.Reminder
import io.github.xalrk.nudge.data.Repeat
import org.json.JSONArray
import org.json.JSONObject
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

data class ParseResult(val reminders: List<Reminder>, val errors: List<String>)

/**
 * Two import formats, auto-detected:
 *
 * 1. Plain text, one reminder per line (easy to write by hand):
 *      Drink some water
 *      Call mom @ 2026-09-10 14:30
 *      Stretch @ 09:00 every day
 *      Team sync @ 2026-09-08 10:00 every mon,wed,fri until 2026-12-19
 *      Pay rent @ 2026-10-01 09:00 every month :: transfer before noon
 *    Lines with no "@" become random reminders. "#" starts a comment.
 *
 * 2. JSON, an array (or {"reminders": [...]}) of objects:
 *      {"title": "Call mom", "body": "", "at": "2026-09-10T14:30", "repeat": "weekly",
 *       "interval": 1, "weekdays": ["mon","wed"], "until": "2026-12-31",
 *       "zone": "America/Denver", "floating": true}
 *    Omit "at" for a random reminder. This is also the export format.
 */
object ImportExport {
    private val dateFmt = DateTimeFormatter.ISO_LOCAL_DATE
    private val dateTimeFmt = Reminder.DT_FORMAT

    fun parse(content: String, now: LocalDateTime = LocalDateTime.now()): ParseResult {
        val trimmed = content.trim()
        return if (trimmed.startsWith("[") || trimmed.startsWith("{")) parseJson(trimmed) else parseText(trimmed, now)
    }

    // ---------------------------------------------------------------- text

    private val dateRe = Regex("""\b(\d{4}-\d{2}-\d{2})\b""")
    private val timeRe = Regex("""\b(\d{1,2}):(\d{2})\s*(am|pm)?\b""", RegexOption.IGNORE_CASE)
    private val everyRe = Regex("""\bevery\s+(?:(\d+)\s+)?([a-z,]+)""", RegexOption.IGNORE_CASE)
    private val untilRe = Regex("""\buntil\s+(\d{4}-\d{2}-\d{2})""", RegexOption.IGNORE_CASE)
    private val clauseStartRe = Regex("""\s(@|every\s|until\s)""", RegexOption.IGNORE_CASE)

    fun parseText(content: String, now: LocalDateTime = LocalDateTime.now()): ParseResult {
        val out = ArrayList<Reminder>()
        val errors = ArrayList<String>()
        content.lines().forEachIndexed { idx, raw ->
            val line = raw.trim()
            if (line.isEmpty() || line.startsWith("#")) return@forEachIndexed
            runCatching { parseLine(line, now) }
                .onSuccess { out += it }
                .onFailure { errors += "Line ${idx + 1}: ${it.message ?: "could not parse"}" }
        }
        return ParseResult(out, errors)
    }

    fun parseLine(line: String, now: LocalDateTime = LocalDateTime.now()): Reminder {
        var main = line
        var body = ""
        val sep = line.indexOf("::")
        if (sep >= 0) { main = line.substring(0, sep).trim(); body = line.substring(sep + 2).trim() }

        val clauseAt = clauseStartRe.find(main)?.range?.first
        val title = (if (clauseAt != null) main.substring(0, clauseAt) else main).trim()
        val clauses = if (clauseAt != null) main.substring(clauseAt).trim() else ""
        require(title.isNotEmpty()) { "missing reminder text" }

        if (clauses.isEmpty() || !clauses.contains("@")) {
            require(clauses.isEmpty()) { "repeat rules need a time, e.g. \"@ 09:00 every day\"" }
            return Reminder(title = title, body = body, kind = Kind.RANDOM).withDedupeKey()
        }

        val date = dateRe.find(clauses)?.groupValues?.get(1)?.let { LocalDate.parse(it) }
        val untilMatch = untilRe.find(clauses)
        val until = untilMatch?.groupValues?.get(1)?.let { LocalDate.parse(it) }
        // The "until" date must not be mistaken for the start date.
        val startDate = if (date != null && untilMatch != null && untilMatch.value.contains(date.toString()))
            dateRe.findAll(clauses).map { LocalDate.parse(it.groupValues[1]) }.firstOrNull { it != until } else date

        val time = timeRe.find(clauses.replace(untilRe, ""))?.let { m ->
            var h = m.groupValues[1].toInt()
            val min = m.groupValues[2].toInt()
            val ampm = m.groupValues[3].lowercase()
            if (ampm == "pm" && h < 12) h += 12
            if (ampm == "am" && h == 12) h = 0
            LocalTime.of(h, min)
        } ?: LocalTime.of(9, 0)

        var repeat = Repeat.NONE
        var interval = 1
        var weekdays = 0
        everyRe.find(clauses)?.let { m ->
            interval = m.groupValues[1].toIntOrNull() ?: 1
            val unit = m.groupValues[2].lowercase()
            when {
                unit.startsWith("day") -> repeat = Repeat.DAILY
                unit.startsWith("week") && !unit.startsWith("weekday") && !unit.startsWith("weekend") -> repeat = Repeat.WEEKLY
                unit.startsWith("month") -> repeat = Repeat.MONTHLY
                unit.startsWith("year") -> repeat = Repeat.YEARLY
                unit.startsWith("weekday") -> { repeat = Repeat.WEEKLY; weekdays = Reminder.maskOf(DayOfWeek.entries.filter { it.value <= 5 }) }
                unit.startsWith("weekend") -> { repeat = Repeat.WEEKLY; weekdays = Reminder.maskOf(listOf(DayOfWeek.SATURDAY, DayOfWeek.SUNDAY)) }
                else -> {
                    val days = unit.split(',').mapNotNull { parseDay(it) }
                    require(days.isNotEmpty()) { "unknown repeat unit \"$unit\"" }
                    repeat = Repeat.WEEKLY; weekdays = Reminder.maskOf(days)
                }
            }
        }

        var start = (startDate ?: now.toLocalDate()).atTime(time)
        if (startDate == null) {
            // Only a time was given: use today, or the next matching day if that moment already passed.
            if (!start.isAfter(now)) start = start.plusDays(1)
            if (repeat == Repeat.WEEKLY && weekdays != 0) {
                val set = DayOfWeek.entries.filter { weekdays and (1 shl (it.value - 1)) != 0 }.toSet()
                var guard = 0
                while (start.dayOfWeek !in set && guard++ < 7) start = start.plusDays(1)
            }
        }

        return Reminder(
            title = title, body = body, kind = Kind.SCHEDULED,
            localDateTime = start.format(dateTimeFmt), zoneId = ZoneId.systemDefault().id, floating = true,
            repeat = repeat, interval = interval.coerceAtLeast(1), weekdays = weekdays,
            endDate = until?.format(dateFmt),
        ).withDedupeKey()
    }

    private fun parseDay(s: String): DayOfWeek? {
        val k = s.trim().lowercase(Locale.ROOT)
        if (k.length < 2) return null
        return DayOfWeek.entries.firstOrNull { it.name.lowercase(Locale.ROOT).startsWith(k) || k.startsWith(it.name.lowercase(Locale.ROOT).take(3)) }
    }

    // ---------------------------------------------------------------- json

    fun parseJson(content: String): ParseResult {
        val out = ArrayList<Reminder>()
        val errors = ArrayList<String>()
        val arr: JSONArray = try {
            if (content.startsWith("[")) JSONArray(content) else JSONObject(content).optJSONArray("reminders") ?: JSONArray()
        } catch (e: Exception) {
            return ParseResult(emptyList(), listOf("Invalid JSON: ${e.message}"))
        }
        for (i in 0 until arr.length()) {
            runCatching { fromJson(arr.getJSONObject(i)) }
                .onSuccess { out += it }
                .onFailure { errors += "Item ${i + 1}: ${it.message ?: "could not parse"}" }
        }
        return ParseResult(out, errors)
    }

    private fun fromJson(o: JSONObject): Reminder {
        val title = o.optString("title").trim().ifEmpty { o.optString("message").trim() }
        require(title.isNotEmpty()) { "missing \"title\"" }
        val body = o.optString("body", "")
        val at = o.optString("at", "").trim()
        val enabled = o.optBoolean("enabled", true)
        if (at.isEmpty()) return Reminder(title = title, body = body, kind = Kind.RANDOM, enabled = enabled).withDedupeKey()

        val ldt = runCatching { LocalDateTime.parse(at) }.getOrElse {
            runCatching { LocalDate.parse(at).atTime(9, 0) }.getOrElse { throw IllegalArgumentException("bad \"at\" value \"$at\" (use 2026-09-10T14:30)") }
        }
        val repeat = when (o.optString("repeat", "none").lowercase()) {
            "", "none", "once" -> Repeat.NONE
            "daily", "day" -> Repeat.DAILY
            "weekly", "week" -> Repeat.WEEKLY
            "monthly", "month" -> Repeat.MONTHLY
            "yearly", "year", "annually" -> Repeat.YEARLY
            else -> throw IllegalArgumentException("unknown repeat \"${o.optString("repeat")}\"")
        }
        val weekdays = o.optJSONArray("weekdays")?.let { a ->
            Reminder.maskOf((0 until a.length()).mapNotNull { parseDay(a.getString(it)) })
        } ?: 0
        val zone = o.optString("zone", "").ifEmpty { ZoneId.systemDefault().id }
        runCatching { ZoneId.of(zone) }.getOrElse { throw IllegalArgumentException("unknown zone \"$zone\"") }
        val until = o.optString("until", "").ifEmpty { null }?.let { LocalDate.parse(it).format(dateFmt) }
        return Reminder(
            title = title, body = body, kind = Kind.SCHEDULED,
            localDateTime = ldt.format(dateTimeFmt), zoneId = zone, floating = o.optBoolean("floating", true),
            repeat = repeat, interval = o.optInt("interval", 1).coerceAtLeast(1), weekdays = weekdays,
            endDate = until, enabled = enabled,
        ).withDedupeKey()
    }

    fun toJson(reminders: List<Reminder>): String {
        val arr = JSONArray()
        for (r in reminders) {
            val o = JSONObject()
            o.put("title", r.title)
            if (r.body.isNotEmpty()) o.put("body", r.body)
            if (r.isScheduled) {
                o.put("at", r.localDateTime)
                o.put("zone", r.zoneId ?: ZoneId.systemDefault().id)
                o.put("floating", r.floating)
                o.put("repeat", r.repeat.name.lowercase())
                if (r.repeat != Repeat.NONE) o.put("interval", r.interval)
                if (r.repeat == Repeat.WEEKLY && r.weekdays != 0) {
                    o.put("weekdays", JSONArray(r.weekdaySet().map { it.name.take(3).lowercase() }))
                }
                r.endDate?.let { o.put("until", it) }
            }
            o.put("enabled", r.enabled)
            arr.put(o)
        }
        return JSONObject().put("reminders", arr).toString(2)
    }
}

File: ios/NudgeCore/Sources/NudgeCore/ImportExport.swift
import Foundation

public struct ParseResult: Equatable {
    public var reminders: [Reminder]
    public var errors: [String]
    public init(reminders: [Reminder], errors: [String]) { self.reminders = reminders; self.errors = errors }
}

struct ImportError: Error, CustomStringConvertible {
    let description: String
    init(_ d: String) { description = d }
}

/// Import/export format: CSV with one reminder per row.
///
///     title,details,date,time,repeat,every,weekdays,until,zone,follow_device_zone
///     Drink some water,,,,,,,,,
///     Call mom,she is home on Sundays,2026-09-14,18:00,weekly,1,sun,,,
///     Standup,,2026-09-07,09:30,weekly,1,mon;tue;wed;thu;fri,2026-12-19,,
///     Flight,,2026-10-20,06:30,,,,,America/Denver,no
///
/// - The header row is optional; without it the columns are read in the order above.
///   With it, columns may appear in any order and unknown columns are ignored.
/// - Only `title` is required. A row with no date and no time is a random reminder.
/// - Weekdays are separated by ";" or spaces (commas work too if the cell is quoted).
///
/// JSON in the same shape ({"reminders": [{"title": ..., "at": "2026-09-10T14:30", ...}]})
/// is still accepted for people who prefer it. Byte-for-byte the same rules as the Android app.
public enum ImportExport {
    public static let csvColumns = ["title", "details", "date", "time", "repeat", "every", "weekdays", "until", "zone", "follow_device_zone"]

    private static let columnAliases: [String: String] = [
        "title": "title", "message": "title", "text": "title", "name": "title", "reminder": "title",
        "details": "details", "body": "details", "description": "details", "notes": "details", "note": "details",
        "date": "date", "day": "date", "start": "date", "start_date": "date",
        "time": "time", "at": "time", "start_time": "time",
        "repeat": "repeat", "recurrence": "repeat", "frequency": "repeat",
        "every": "every", "interval": "every",
        "weekdays": "weekdays", "days": "weekdays", "on": "weekdays",
        "until": "until", "end": "until", "end_date": "until", "ends": "until",
        "zone": "zone", "timezone": "zone", "time_zone": "zone", "tz": "zone",
        "follow_device_zone": "follow_device_zone", "floating": "follow_device_zone", "follow_device_timezone": "follow_device_zone",
    ]

    private static let bom = "\u{FEFF}"

    public static func parse(_ content: String, now: LocalDateTime = Date().toLocal(in: .device), zone: TimeZone = .device) -> ParseResult {
        var trimmed = content.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix(bom) { trimmed.removeFirst() }
        return (trimmed.hasPrefix("[") || trimmed.hasPrefix("{")) ? parseJson(trimmed, zone: zone) : parseCsv(trimmed, now: now, zone: zone)
    }

    // ----------------------------------------------------------------- csv

    public static func parseCsv(_ content0: String, now: LocalDateTime = Date().toLocal(in: .device), zone: TimeZone = .device) -> ParseResult {
        var content = content0
        if content.hasPrefix(bom) { content.removeFirst() }
        let rows = readCsv(content)
        var out: [Reminder] = []
        var errors: [String] = []
        if rows.isEmpty { return ParseResult(reminders: out, errors: errors) }

        let first = rows[0]
        let hasHeader = (!first.isEmpty && columnAliases[normHeader(first[0])] == "title" && first.count > 1) ||
            (!first.isEmpty && first.allSatisfy { columnAliases[normHeader($0)] != nil || $0.isBlank } && first.contains { !$0.isBlank } && first.count > 1)
        let columns: [String?] = hasHeader ? first.map { columnAliases[normHeader($0)] } : csvColumns

        for (i, cells) in rows.dropFirst(hasHeader ? 1 : 0).enumerated() {
            let lineNo = i + 1 + (hasHeader ? 1 : 0)
            if cells.allSatisfy({ $0.isBlank }) { continue }
            if let f = cells.first, f.drop(while: { $0.isWhitespace }).hasPrefix("#") { continue }
            var field: [String: String] = [:]
            for (c, name) in columns.enumerated() {
                if let name = name, c < cells.count { field[name] = cells[c].trimmingCharacters(in: .whitespaces) }
            }
            do { out.append(try fromFields(field, now: now, zone: zone)) }
            catch { errors.append("Row \(lineNo): \(String(describing: error))") }
        }
        return ParseResult(reminders: out, errors: errors)
    }

    private static func normHeader(_ s: String) -> String {
        var t = s.trimmingCharacters(in: .whitespaces).lowercased()
        // Runs of whitespace or dashes become a single underscore.
        var outS = ""
        var inRun = false
        for ch in t {
            if ch.isWhitespace || ch == "-" {
                if !inRun { outS.append("_"); inRun = true }
            } else { outS.append(ch); inRun = false }
        }
        t = outS
        if t.hasPrefix(bom) { t.removeFirst() }
        return t
    }

    /// Minimal RFC 4180 reader: quoted cells, doubled quotes, newlines inside quotes. Auto-detects ";" delimiters.
    /// Works on unicode scalars because Swift treats "\r\n" as a single Character.
    public static func readCsv(_ text: String) -> [[String]] {
        let scalars = Array(text.unicodeScalars)
        var firstLine = ""
        for u in scalars { if u == "\n" { break }; firstLine.unicodeScalars.append(u) }
        let delim: Unicode.Scalar = (!firstLine.contains(",") && firstLine.contains(";")) ? ";" : ","
        var rows: [[String]] = []
        var row: [String] = []
        var cell = ""
        var inQuotes = false
        var i = 0
        while i < scalars.count {
            let ch = scalars[i]
            if inQuotes {
                if ch == "\"" {
                    if i + 1 < scalars.count && scalars[i + 1] == "\"" { cell.append("\""); i += 1 } else { inQuotes = false }
                } else { cell.unicodeScalars.append(ch) }
            } else if ch == "\"" { inQuotes = true }
            else if ch == delim { row.append(cell); cell = "" }
            else if ch == "\r" { }
            else if ch == "\n" { row.append(cell); cell = ""; rows.append(row); row = [] }
            else { cell.unicodeScalars.append(ch) }
            i += 1
        }
        if !cell.isEmpty || !row.isEmpty { row.append(cell); rows.append(row) }
        return rows
    }

    private static func fromFields(_ f: [String: String], now: LocalDateTime, zone: TimeZone) throws -> Reminder {
        let title = (f["title"] ?? "").trimmingCharacters(in: .whitespaces)
        guard !title.isEmpty else { throw ImportError("missing title") }
        let body = (f["details"] ?? "").trimmingCharacters(in: .whitespaces)
        let dateStr = f["date"] ?? ""
        let timeStr = f["time"] ?? ""
        if dateStr.isBlank && timeStr.isBlank {
            guard (f["repeat"] ?? "").isBlank else { throw ImportError("a repeating reminder needs a date or time") }
            return Reminder(title: title, body: body, kind: .random).withDedupeKey()
        }

        let date = dateStr.isBlank ? nil : try parseDate(dateStr)
        let time = timeStr.isBlank ? LocalTime(9, 0) : try parseTime(timeStr)
        let (rep, shorthandDays) = try parseRepeat(f["repeat"] ?? "")
        var weekdays = shorthandDays
        if let s = f["weekdays"], !s.isBlank {
            let parts = s.split(whereSeparator: { ";,/|".contains($0) || $0.isWhitespace }).map(String.init).filter { !$0.isBlank }
            var days: [DayOfWeek] = []
            for p in parts {
                guard let d = parseDay(p) else { throw ImportError("unknown weekday \"\(p)\"") }
                days.append(d)
            }
            weekdays = Reminder.maskOf(days)
        }
        var every = 1
        if let e = f["every"], !e.isBlank {
            guard let n = Int(e.trimmingCharacters(in: .whitespaces)) else { throw ImportError("\"every\" must be a number") }
            every = n
        }
        let until: LocalDate? = (f["until"] ?? "").isBlank ? nil : try parseDate(f["until"]!)
        var tz = zone
        if let z = f["zone"], !z.isBlank {
            guard let t = TimeZone(identifier: z.trimmingCharacters(in: .whitespaces)) else { throw ImportError("unknown zone \"\(z)\"") }
            tz = t
        }
        let floating = (f["follow_device_zone"] ?? "").isBlank ? true : try parseBool(f["follow_device_zone"]!)

        let effectiveRepeat: Repeat = (rep == .none && weekdays != 0) ? .weekly : rep
        var start = (date ?? now.date).atTime(time)
        if date == nil {
            if !start.isAfter(now) { start = start.plusDays(1) }
            if effectiveRepeat == .weekly && weekdays != 0 {
                let set = DayOfWeek.set(fromMask: weekdays)
                var guardCount = 0
                while !set.contains(start.dayOfWeek) && guardCount < 7 { start = start.plusDays(1); guardCount += 1 }
            }
        }
        return Reminder(
            title: title, body: body, kind: .scheduled,
            localDateTime: start.description, zoneId: tz.identifier, floating: floating,
            repeatRule: effectiveRepeat, interval: max(every, 1),
            weekdays: effectiveRepeat == .weekly ? weekdays : 0,
            endDate: until?.description
        ).withDedupeKey()
    }

    private static let monthNames = ["jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"]

    static func parseDate(_ s: String) throws -> LocalDate {
        let t = s.trimmingCharacters(in: .whitespaces)
        if let d = LocalDate(iso: t) { return d }
        func numeric(_ parts: [String], _ order: (Int, Int, Int)) -> LocalDate? {
            guard parts.count == 3, let a = Int(parts[order.0]), let b = Int(parts[order.1]), let c = Int(parts[order.2]) else { return nil }
            return LocalDate.isValid(a, b, c) ? LocalDate(a, b, c) : nil
        }
        // yyyy/MM/dd, yyyy.MM.dd (two-digit month and day), then d/M/yyyy, M/d/yyyy, d.M.yyyy.
        for sep in ["/", "."] {
            let p = t.split(separator: Character(sep), omittingEmptySubsequences: false).map(String.init)
            if p.count == 3 {
                if p[0].count == 4 && p[1].count == 2 && p[2].count == 2, let d = numeric(p, (0, 1, 2)) { return d }
                if p[2].count == 4 {
                    if let d = numeric(p, (2, 1, 0)) { return d }          // d/M/yyyy
                    if sep == "/", let d = numeric(p, (2, 0, 1)) { return d } // M/d/yyyy
                }
            }
        }
        // "MMM d yyyy" and "d MMM yyyy" with English month abbreviations.
        let words = t.split(whereSeparator: { $0.isWhitespace }).map(String.init)
        if words.count == 3 {
            func month(_ w: String) -> Int? { monthNames.firstIndex(of: String(w.lowercased().prefix(3))).map { $0 + 1 } }
            if let m = month(words[0]), let d = Int(words[1]), let y = Int(words[2]), words[2].count == 4, LocalDate.isValid(y, m, d) { return LocalDate(y, m, d) }
            if let d = Int(words[0]), let m = month(words[1]), let y = Int(words[2]), words[2].count == 4, LocalDate.isValid(y, m, d) { return LocalDate(y, m, d) }
        }
        throw ImportError("bad date \"\(s)\" (use YYYY-MM-DD)")
    }

    static func parseTime(_ s: String) throws -> LocalTime {
        var t = s.trimmingCharacters(in: .whitespaces).lowercased()
        var ap = ""
        for suffix in ["am", "pm", "a", "p"] where t.hasSuffix(suffix) {
            ap = String(suffix.prefix(1))
            t = String(t.dropLast(suffix.count)).trimmingCharacters(in: .whitespaces)
            break
        }
        let parts = t.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 1 || parts.count == 2, (1...2).contains(parts[0].count), parts[0].allSatisfy({ $0.isNumber }),
              var h = Int(parts[0]) else { throw ImportError("bad time \"\(s)\" (use HH:MM)") }
        var min = 0
        if parts.count == 2 {
            guard parts[1].count == 2, parts[1].allSatisfy({ $0.isNumber }), let m = Int(parts[1]) else { throw ImportError("bad time \"\(s)\" (use HH:MM)") }
            min = m
        }
        if ap == "p" && h < 12 { h += 12 }
        if ap == "a" && h == 12 { h = 0 }
        guard (0...23).contains(h), (0...59).contains(min) else { throw ImportError("bad time \"\(s)\"") }
        return LocalTime(h, min)
    }

    /// Returns repeat + a weekday mask for shorthand values such as "weekdays".
    static func parseRepeat(_ s: String) throws -> (Repeat, Int) {
        switch s.trimmingCharacters(in: .whitespaces).lowercased() {
        case "", "none", "no", "once", "never", "0": return (.none, 0)
        case "daily", "day", "days", "every day": return (.daily, 0)
        case "weekly", "week", "weeks", "every week": return (.weekly, 0)
        case "weekday", "weekdays": return (.weekly, Reminder.maskOf(DayOfWeek.allCases.filter { $0.rawValue <= 5 }))
        case "weekend", "weekends": return (.weekly, Reminder.maskOf([DayOfWeek.saturday, .sunday]))
        case "monthly", "month", "months", "every month": return (.monthly, 0)
        case "yearly", "year", "years", "annually", "annual", "every year": return (.yearly, 0)
        default: throw ImportError("unknown repeat \"\(s)\" (use daily, weekly, monthly, yearly)")
        }
    }

    private static func parseBool(_ s: String) throws -> Bool {
        switch s.trimmingCharacters(in: .whitespaces).lowercased() {
        case "yes", "y", "true", "1", "on": return true
        case "no", "n", "false", "0", "off": return false
        default: throw ImportError("\"follow_device_zone\" must be yes or no")
        }
    }

    static func parseDay(_ s: String) -> DayOfWeek? {
        let k = s.trimmingCharacters(in: .whitespaces).lowercased()
        if k.count < 2 { return nil }
        return DayOfWeek.allCases.first { d in
            let n = d.name.lowercased()
            return n.hasPrefix(k) || k.hasPrefix(String(n.prefix(3)))
        }
    }

    public static func toCsv(_ reminders: [Reminder]) -> String {
        var s = csvColumns.joined(separator: ",") + "\r\n"
        for r in reminders {
            let ldt = r.localDateTimeOrNil()
            let repeating = r.isScheduled && r.repeatRule != .none
            let cells = [
                r.title,
                r.body,
                ldt.map { $0.date.description } ?? "",
                ldt.map { $0.time.hhmm } ?? "",
                repeating ? r.repeatRule.rawValue.lowercased() : "",
                repeating ? String(r.interval) : "",
                (r.isScheduled && r.repeatRule == .weekly && r.weekdays != 0) ? r.weekdaySet().sorted().map { $0.name.prefix(3).lowercased() }.joined(separator: ";") : "",
                r.endDate ?? "",
                r.isScheduled ? (r.zoneId ?? "") : "",
                r.isScheduled ? (r.floating ? "yes" : "no") : "",
            ]
            s += cells.map(csvCell).joined(separator: ",") + "\r\n"
        }
        return s
    }

    private static func csvCell(_ s: String) -> String {
        if s.contains(where: { $0 == "," || $0 == "\"" || $0 == "\n" || $0 == "\r" || $0 == ";" }) || s.hasPrefix(" ") || s.hasSuffix(" ") {
            return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
        }
        return s
    }

    // ---------------------------------------------------------------- json

    public static func parseJson(_ content: String, zone: TimeZone = .device) -> ParseResult {
        var out: [Reminder] = []
        var errors: [String] = []
        let arr: [Any]
        do {
            let obj = try JSONSerialization.jsonObject(with: Data(content.utf8))
            if let a = obj as? [Any] { arr = a }
            else if let o = obj as? [String: Any] { arr = (o["reminders"] as? [Any]) ?? [] }
            else { arr = [] }
        } catch {
            return ParseResult(reminders: [], errors: ["Invalid JSON: \(error.localizedDescription)"])
        }
        for (i, item) in arr.enumerated() {
            do {
                guard let o = item as? [String: Any] else { throw ImportError("not an object") }
                out.append(try fromJson(o, zone: zone))
            } catch { errors.append("Item \(i + 1): \(String(describing: error))") }
        }
        return ParseResult(reminders: out, errors: errors)
    }

    private static func optString(_ o: [String: Any], _ key: String, _ def: String = "") -> String {
        guard let v = o[key] else { return def }
        if v is NSNull { return def }
        if let s = v as? String { return s }
        if let b = v as? Bool { return b ? "true" : "false" }
        return "\(v)"
    }
    private static func optBool(_ o: [String: Any], _ key: String, _ def: Bool) -> Bool {
        if let b = o[key] as? Bool { return b }
        if let s = o[key] as? String { if s.lowercased() == "true" { return true }; if s.lowercased() == "false" { return false } }
        return def
    }
    private static func optInt(_ o: [String: Any], _ key: String, _ def: Int) -> Int {
        if let n = o[key] as? Int { return n }
        if let d = o[key] as? Double { return Int(d) }
        if let s = o[key] as? String, let n = Int(s) { return n }
        return def
    }

    private static func fromJson(_ o: [String: Any], zone: TimeZone) throws -> Reminder {
        var title = optString(o, "title").trimmingCharacters(in: .whitespaces)
        if title.isEmpty { title = optString(o, "message").trimmingCharacters(in: .whitespaces) }
        guard !title.isEmpty else { throw ImportError("missing \"title\"") }
        let body = optString(o, "body", optString(o, "details", ""))
        let at = optString(o, "at", "").trimmingCharacters(in: .whitespaces)
        let enabled = optBool(o, "enabled", true)
        if at.isEmpty { return Reminder(title: title, body: body, kind: .random, enabled: enabled).withDedupeKey() }

        let ldt: LocalDateTime
        if let l = LocalDateTime(iso: at) { ldt = l }
        else if let d = LocalDate(iso: at) { ldt = d.atTime(9, 0) }
        else { throw ImportError("bad \"at\" value \"\(at)\" (use 2026-09-10T14:30)") }
        let (rep, shorthandDays) = try parseRepeat(optString(o, "repeat", "none"))
        var weekdays = shorthandDays
        if let a = o["weekdays"] as? [Any] {
            weekdays = Reminder.maskOf(a.compactMap { parseDay("\($0)") })
        }
        var zoneId = optString(o, "zone", "")
        if zoneId.isEmpty { zoneId = zone.identifier }
        guard TimeZone(identifier: zoneId) != nil else { throw ImportError("unknown zone \"\(zoneId)\"") }
        var until: String? = nil
        let u = optString(o, "until", "")
        if !u.isEmpty {
            guard let d = LocalDate(iso: u) else { throw ImportError("bad \"until\" value \"\(u)\"") }
            until = d.description
        }
        return Reminder(
            title: title, body: body, kind: .scheduled,
            localDateTime: ldt.description, zoneId: zoneId, floating: optBool(o, "floating", true),
            repeatRule: rep, interval: max(optInt(o, "interval", 1), 1), weekdays: weekdays,
            endDate: until, enabled: enabled
        ).withDedupeKey()
    }
}

public extension String {
    public var isBlank: Bool { allSatisfy { $0.isWhitespace } }
}
